examples/lock-app/linux/main.cpp

```diff
@@ -17,8 +17,47 @@
  */
 
 #include "AppMain.h"
+#include <app/clusters/network-commissioning/network-commissioning.h>
+#include <platform/Linux/NetworkCommissioningDriver.h>
 
-void ApplicationInit() {}
+using namespace chip;
+using namespace chip::app;
+
+namespace {
+#ifdef HAVE_NETWORK_COMMISSIONING_DRIVER
+#undef HAVE_NETWORK_COMMISSIONING_DRIVER
```

Comment:
under what circumstances is this set and what is the difference from code before?

New code sets this define if THREAD or WPA is set, old code just had it from somewhere. Does it make a difference? A comment of why this is undef first may help, specifically static why it was set to start with and why we override.

Reply:
Also do we need to use this `#define`? How about having a `void InitNetwork` that is with a right content including empty if none of thread/wpa are set and have ApplicationInit call it without an ifdef? Less split of ifdefs (and less ifdefs) make code easier to read.

Reply:
Old code never had it.  This is just guarding against someone outside this code altogether setting the define...

Would you prefer that we just #error if it's defined by someone outside this file?

Reply:
Yeah, I can do that.

Reply:
```cpp
// We support either thread or wi-fi commissioning, not both.  Prefer thread if
// both are available.

#if !(CHIP_DEVICE_LAYER_TARGET_LINUX && CHIP_DEVICE_CONFIG_ENABLE_THREAD) && !CHIP_DEVICE_CONFIG_ENABLE_WPA

void NetworkInit() {
    ChipLogWarning("No network provisioning support enabled");
}

#else // !(CHIP_DEVICE_LAYER_TARGET_LINUX && CHIP_DEVICE_CONFIG_ENABLE_THREAD) && !CHIP_DEVICE_CONFIG_ENABLE_WPA

#if (CHIP_DEVICE_LAYER_TARGET_LINUX && CHIP_DEVICE_CONFIG_ENABLE_THREAD)
    using NetworkCommissioningDriver = DeviceLayer::NetworkCommissioning::LinuxThreadDriver;
#elif CHIP_DEVICE_CONFIG_ENABLE_WPA
    using NetworkCommissioningDriver = DeviceLayer::NetworkCommissioning::LinuxWiFiDriver;
#else
    #error "No suitable network driver found"
#endif

NetworkCommissioningDriver sNetworkCommissioningDriver;
Clusters::NetworkCommissioning::Instance sNetworkCommissioningInstance(0, &sNetworkCommissioningDriver);

void NetworkInit() {
    sNetworkCommissioningInstance.Init();
}

#endif // !(CHIP_DEVICE_LAYER_TARGET_LINUX && CHIP_DEVICE_CONFIG_ENABLE_THREAD) && !CHIP_DEVICE_CONFIG_ENABLE_WPA

} // anonymous namespace

void ApplicationInit()
{
    NetworkInit();
}
```

maybe something like this. Ideally we would have a NULLDriver for networking but implementing that is probably non-trivial.

Reply:
```cpp
namespace {

// We support either thread or wi-fi commissioning, not both.  Prefer thread if
// both are available.
#if (CHIP_DEVICE_LAYER_TARGET_LINUX && CHIP_DEVICE_CONFIG_ENABLE_THREAD)
    using NetworkCommissioningDriver = DeviceLayer::NetworkCommissioning::LinuxThreadDriver;
#elif CHIP_DEVICE_CONFIG_ENABLE_WPA
    using NetworkCommissioningDriver = DeviceLayer::NetworkCommissioning::LinuxWiFiDriver;
#else
    using NetworkCommissioningDriver = app::Clusters::NetworkCommissioning::NullNetworkDriver;
#endif

NetworkCommissioningDriver sNetworkCommissioningDriver;
Clusters::NetworkCommissioning::Instance sNetworkCommissioningInstance(0, &sNetworkCommissioningDriver);

} // anonymous namespace

void ApplicationInit()
{
    sNetworkCommissioningInstance.Init();
}

```

assuming we always want some real-looking network driver. Otherwise the more complex code above or some variation like that.